keel-web/src/test/kotlin/com/netflix/spinnaker/keel/rest/ExceptionHandlerTests.kt

```diff
@@ -0,0 +1,309 @@
+            }
+          }
+        }
+      }
+    }
+
+    context("delivery config with invalid top-level field") {
+      fixture {
+        Fixture(
+          """
+            ---
+            name: fnord
+            application: fnord
+            serviceAccount: keel@netlix.com
+            artifacts: true # wrong
+            environments:
+            - name: test
+              constraints: []
+              notifications: []
+              resources:
+              - apiVersion: $TEST_API
+                kind: "whatever"
+                spec: {}
+          """.trimIndent()
+        )
+      }
+
+      test("should cause an ApiError with the expected details") {
+        expect {
+          that(apiError.details) {
+            isA<ParsingErrorDetails>().and {
+              get { error }.isEqualTo(ParsingError.INVALID_TYPE)
+              get { path }.size.isEqualTo(1)
+              get { pathExpression }.isEqualTo(".artifacts")
+            }
+          }
+        }
+      }
+    }
+
+    context("environment with missing required field") {
+      fixture {
+        Fixture(
+          """
+            ---
+            name: fnord
+            application: fnord
+            serviceAccount: keel@netlix.com
+            artifacts: []
+            environments:
+            - constraints: []
+              notifications: []
+              resources:
+              - apiVersion: $TEST_API
+                kind: "whatever"
+                spec: {}
+          """.trimIndent()
+        )
+      }
+
+      test("should cause an ApiError with the expected details") {
+        expect {
+          that(apiError.details) {
+            isA<ParsingErrorDetails>().and {
+              get { error }.isEqualTo(ParsingError.INVALID_VALUE)
+              get { path }.size.isEqualTo(2)
+              get { pathExpression }.isEqualTo(".environments[0]")
+            }
+          }
+        }
+      }
+    }
+
+    context("environment with invalid type for field") {
+      fixture {
+        Fixture(
+          """
+            ---
+            name: fnord
+            application: fnord
+            serviceAccount: keel@netlix.com
+            artifacts: []
+            environments:
+            - name: test
+              constraints: true # wrong
+              notifications: []
+              resources:
+              - apiVersion: $TEST_API
+                kind: "whatever"
+                spec: {}
+          """.trimIndent()
+        )
+      }
+
+      test("should cause an ApiError with the expected details") {
+        expect {
+          that(apiError.details) {
+            isA<ParsingErrorDetails>().and {
+              get { error }.isEqualTo(ParsingError.INVALID_TYPE)
+              get { path }.size.isEqualTo(3)
+              get { pathExpression }.isEqualTo(".environments[0].constraints")
+            }
+          }
+        }
+      }
+    }
+
+    context("resource missing field checked by validation") {
```

Comment:
This context name makes a good point, can you add a comment somewhere here or in a github issue documenting the gaps in getting the new `/delivery-configs/validate` endpoint to fully validate resources?

Reply:
Thanks for the feedback, @asher. I feel like most of this will go away when @robfletcher is done with the JSON schema stuff. I expect we'll replace my poor-man's validation with a check against the schema. Is your point that, even then, there might be validations that are _not_ part of the model definition and might need to be checked on the side?

Reply:
Complete resource validation has to happen after resolvers are applied but with that in place, json schema validation should be sufficient.

Reply:
OK. I already had a `TODO` in there to replace the `ApiError` validation with a check using the schema.